[ONNX] Enable export of __xor_ (#64042)

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#64581 [ONNX] Enable export of __xor_ (#64042)**
* #64580 [ONNX] Fix inplace fill_ dtype export mismatch (#64233)
* #64579 [ONNX] Update doc and error message for indexing export (#64290)
* #64578 [ONNX] Fix remainder export (#64230)
* #64381 Add onnx test for batched_nms (#53175)
* #64379 [ONNX] Fix gather squeeze axis in constant folding (#63588)
* #64378 [ONNX] Fix cuda test case (#63597)
* #64377 [ONNX] Fix input sequence for pad op (#60554)
* #64376 Fix empty size constant creation (#63607)
* #64375 [ONNX] Update instance_norm implementation and support training (#60538)

* Enbale xor

* Update test_pytorch_onnx_onnxruntime.py

* Update symbolic_opset9.py

* Update symbolic_opset9.py

* Update test_pytorch_onnx_onnxruntime.py

* Update symbolic_opset9.py

Differential Revision: [D30919598](https://our.internmc.facebook.com/intern/diff/D30919598)

neginraoof committed:
[ONNX] Enable export of __xor_ (#64042)

* Enbale xor

* Update test_pytorch_onnx_onnxruntime.py

* Update symbolic_opset9.py

* Update symbolic_opset9.py

* Update test_pytorch_onnx_onnxruntime.py

* Update symbolic_opset9.py

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Enable export of __xor_ (#64042)"


* Enbale xor

* Update test_pytorch_onnx_onnxruntime.py

* Update symbolic_opset9.py

* Update symbolic_opset9.py

* Update test_pytorch_onnx_onnxruntime.py

* Update symbolic_opset9.py

[ghstack-poisoned]

msaroufim commented:
@msaroufim has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30919598).

malfet commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -1084,10 +1084,11 @@ def __interpolate(g, input, size, scale_factor, mode , align_corners, recompute_
 def bitwise_not(g, inp):
     if inp.type().scalarType() != "Bool":
-        return _unimplemented("bitwise_not", "non-bool tensor")
+        raise NotImplementedError("ONNX export does NOT support exporting bitwise Not " +
+                                  "for non-boolean input values")
```
String literals are automatically concatenated https://docs.python.org/3/reference/lexical_analysis.html#string-literal-concatenation
```suggestion
        raise NotImplementedError("ONNX export does NOT support exporting bitwise Not "
                                  "for non-boolean input values")
```

malfet commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -1116,6 +1117,9 @@ def wrap_with_not(g, input, other):
 
 def __not_(g, self):
+    if self.type().scalarType() != "Bool":
+        raise NotImplementedError("ONNX export does NOT support exporting bitwise Not " +
+                                  "for non-boolean input values")
```
Same as above
```suggestion
        raise NotImplementedError("ONNX export does NOT support exporting bitwise Not "
                                  "for non-boolean input values")
```

malfet commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -1162,14 +1166,31 @@ def le(g, input, other):
+            other.type().scalarType() == "Bool":
+        return g.op("And", input, other)
+    else:
+        raise NotImplementedError("ONNX export does NOT support exporting bitwise AND " +
+                                  "for non-boolean input values")
```
Same as above
```suggestion
        raise NotImplementedError("ONNX export does NOT support exporting bitwise AND "
                                  "for non-boolean input values")
```

malfet commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -1162,14 +1166,31 @@ def le(g, input, other):
+            other.type().scalarType() == "Bool":
+        return g.op("Or", input, other)
+    else:
+        raise NotImplementedError("ONNX export does NOT support exporting bitwise OR " +
+                                  "for non-boolean input values")
```
Same as above
```suggestion
        raise NotImplementedError("ONNX export does NOT support exporting bitwise OR "
                                  "for non-boolean input values")
```

malfet commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -1162,14 +1166,31 @@ def le(g, input, other):
+            other.type().scalarType() == "Bool":
+        return g.op("Xor", input, other)
+    else:
+        raise NotImplementedError("ONNX export does NOT support exporting bitwise XOR " +
+                                  "for non-boolean input values")
```
Same as above
```suggestion
        raise NotImplementedError("ONNX export does NOT support exporting bitwise XOR "
                                  "for non-boolean input values")
```

BowenBao committed:
Update on "[ONNX] Enable export of __xor_ (#64042)"


* Enbale xor

* Update test_pytorch_onnx_onnxruntime.py

* Update symbolic_opset9.py

* Update symbolic_opset9.py

* Update test_pytorch_onnx_onnxruntime.py

* Update symbolic_opset9.py

Differential Revision: [D30919598](https://our.internmc.facebook.com/intern/diff/D30919598)

[ghstack-poisoned]

msaroufim commented:
@msaroufim has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30919598).

BowenBao committed:
Update on "[ONNX] Enable export of __xor_ (#64042)"


* Enbale xor

* Update test_pytorch_onnx_onnxruntime.py

* Update symbolic_opset9.py

* Update symbolic_opset9.py

* Update test_pytorch_onnx_onnxruntime.py

* Update symbolic_opset9.py

Differential Revision: [D30919598](https://our.internmc.facebook.com/intern/diff/D30919598)

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Enable export of __xor_ (#64042)"


* Enbale xor

* Update test_pytorch_onnx_onnxruntime.py

* Update symbolic_opset9.py

* Update symbolic_opset9.py

* Update test_pytorch_onnx_onnxruntime.py

* Update symbolic_opset9.py

Differential Revision: [D30919598](https://our.internmc.facebook.com/intern/diff/D30919598)

[ghstack-poisoned]

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30919598).

BowenBao committed:
Update on "[ONNX] Enable export of __xor_ (#64042)"


* Enbale xor

* Update test_pytorch_onnx_onnxruntime.py

* Update symbolic_opset9.py

* Update symbolic_opset9.py

* Update test_pytorch_onnx_onnxruntime.py

* Update symbolic_opset9.py

Differential Revision: [D30919598](https://our.internmc.facebook.com/intern/diff/D30919598)

[ghstack-poisoned]

BowenBao commented:
@malfet rebased, please import.

BowenBao committed:
Update on "[ONNX] Enable export of __xor_ (#64042)"


* Enbale xor

* Update test_pytorch_onnx_onnxruntime.py

* Update symbolic_opset9.py

* Update symbolic_opset9.py

* Update test_pytorch_onnx_onnxruntime.py

* Update symbolic_opset9.py

Differential Revision: [D30919598](https://our.internmc.facebook.com/intern/diff/D30919598)

[ghstack-poisoned]

malfet commented:
@BowenBao Not going to reimport in that case
